_episodes/l3_03_test_fixture.md

```diff
@@ -0,0 +1,505 @@
+    calcNetSalary();
+
+}
+
+// Definition of function setName
+void Employee::setName(const string& employee_name) 
+{
+    if(employee_name == "")
+    {
+        throw invalid_argument("Name cannot be empty");
+    }
+    name = employee_name;
+}
+
+// Employee gets additional £1000 bonus if they have been employed for more than 10 years.
+void Employee::calcNetBonus() 
+{   
+    if(number_years_employed > years_exp_for_extra_bonus)
+    {
+        net_bonus = basic_bonus + extra_bonus;
+    }
+    else
+    {
+        net_bonus = basic_bonus;
+    }
+}
+
+// Function to calculate tex amount.
+void Employee::calcTaxAmount() 
+{
+    double salary_with_bonus = base_salary + net_bonus;
+    if(salary_with_bonus <= 10000)
+    {
+        tax_amount = 0.0;
+    }
+    else if(salary_with_bonus > 10000 && salary_with_bonus <= 20000)
+    {
+        tax_amount = 0.1*(salary_with_bonus-10000);
+    }
+    else if(salary_with_bonus > 20000 && salary_with_bonus <= 50000)
+    {
+        tax_amount = 0.1*10000 + 0.2*(salary_with_bonus-20000);
+    }
+    else if(salary_with_bonus > 50000)
+    {
+        tax_amount = 0.1*10000 + \
+                     0.2*30000 + \
+                     0.5*(salary_with_bonus-50000);
+    }
+}
+
+// Function to calculate the ney salary.
+void Employee::calcNetSalary() 
+{
+    net_salary = base_salary + net_bonus - tax_amount;
+}
+```
+With this code, we now have the necessary fragments to test our `Employee` class. Let us see this in action.
+
+## 3. Unit tests for our employee class without test fixture
+
+In order to clearly demonstrate why a test fixture would be needed, we first write some tests for our employee class without using a fixture. This will help us to understand why does a fixture is useful and how to use it.
+
+For this subsection, let us assume that we are checking two functionalities of our employee class (for code, please see [1_employeetest.cpp](../code/Chapter3/1_employeetest.cpp)) which are:-
+
+1. We can set the name of employee correctly.
+2. We can set the age correctly.
+
+The code for these two tests (which is entirely based on what we learn in second module of this course) is given below.
+
+```cpp
+// Test if we can set the name of an employee.
+TEST(EmployeeTest, CanSetName) {
+    Employee employee{"John", 25, 10000, 5, 1000};
+    employee.setName("John Doe");
+    EXPECT_EQ(employee.getName(), "John Doe");
+}
+
+// Test if we can set the age of an employee.
+TEST(EmployeeTest, CanSetAge) {
+    Employee employee{"John", 25, 10000, 5, 1000};
+    employee.setAge(30);
+    EXPECT_EQ(employee.getAge(), 30);
+}
+```
+
+While the above test solve our problem, there is a problem of code duplication and object creation for each test. As we can see in each test that we have to create an instance of employee class by using the statement `Employee employee{"John", 25, 10000, 5, 1000};`. This is against the `DRY (Don't Repeat Yourself)' https://en.wikipedia.org/wiki/Don%27t_repeat_yourself.
```

Comment:
I would try to drive the motivation behind this even further by noting that the setup (and teardown?) function itself could potentially be quite complex -- e.g. loading custom configuration files or connecting to a test SQL server. I probably wouldn't be so convinced by the argument if all I had to do is construct a simple object each time.

Reply:
@jamesturner246 Although I understood your point, I am not sure that I can integrate it well here. The text you have highlighted here was a transition from a normal test (without fixtures) to use of fixtures. We have not yet discussed anything about the Setup and Teardown yet. That is why, I kept the explanation like that. If you want, you can create an issue for this and add what you would like to add here later.